WordPress/openverse-catalog
# Use Airflow base Docker image

## Fixes


Fixes #532 by @AetherUnbound

## Description


This PR replaces our current custom Airflow docker image with the official upstream Airflow docker image. With this comes a number of complications and changes. I don't see this as a solid "win". There are lots of pros and cons that are worth weighing before we decide to merge (or not) this change:

**Pros**:
- Running database migrations on startup is baked into the image for us and just requires setting an environment variable.
- Creating a web UI admin user is as easy as setting a few variables, we no longer need to add logic for this.
- We can remove the bespoke `wait_for_db.py` script since [waiting for the database to be ready is baked in](https://airflow.apache.org/docs/docker-stack/entrypoint.html#waits-for-airflow-db-connection).
- The main processes no longer runs as the `root` user, which [is more secure](https://docs.bitnami.com/tutorials/why-non-root-containers-are-important-for-security) and follows the principle of least privilege.
- Reduced maintenance burden - the Airflow folks put in a lot of effort to make sure that the Docker images are production ready.
- We would have separate containers for the webserver & scheduler (rather than having them run on the same container), which helps isolate those services more.

**Cons**:
- Running as a non-root user can cause complications with bind-mounted volumes. Permissions on the host machine will no longer match up with the Docker user, so certain operations have to be adjusted. This can be seen in changes to the `justfile` (specifically the `generate-dag-docs` recipe, which can no longer write out the file without special considerations) or to tests (see the changes to `test_create_provider_ingester.py`). This is not necessarily a bad thing, but it has brought me a bit of frustration and can cause trouble and strange behavior down the line.
- Startup time is slower. Since we're now handing over management of the migrations to Airflow and the upstream entrypoint has [lots of other startup behavior](https://airflow.apache.org/docs/docker-stack/entrypoint.html), there's more that runs at each container start. In production this is not a problem (since the containers shouldn't be starting & stopping regularly), but it is noticeable in development. Many of our `just` recipes take a few seconds longer to run. Nothing insurmountable from my perspective, just something I've noticed.
- Held back on Python releases. We were able to jump on Python 3.10 before it was officially supported by Airflow. This was because Airflow has a large number of provider libraries which it also supports, and it needed to ensure compatibility with all of those libraries before the application as a whole was ready for Python 3.10. We use a very small portion of the available providers, so as it stands we are held back from upgrading to 3.11 until those upstream issues with providers we don't use are resolved (see #855).
- We now need to set up a CSRF secret key for **both development and production environments** (they are currently defined in neither).

**Other changes**:
- Airflow home folder got changed from `/usr/local/airflow` to `/opt airflow`
- I fixed an issue with the create provider ingester tests wherein we were accidentally handing in a string where a list should go and it was producing one provider for each letter of the word `image`.

## Testing Instructions

Run `just build` on this branch. Make sure you're able to run every `just` recipe we have without error. You may need to run `just recreate` if you encounter any errors initially.

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
